### ISSUE TITLE
InstantUpload syncs InstantUpload folders to wrong account/server.

### ISSUE THREAD

**Issue by eyduh** [nextcloud/android#11645]
### ⚠️ Before posting ⚠️

- [X] This is a **bug**, not a question or an enhancement.
- [X] I've [searched for similar issues](https://github.com/nextcloud/android/issues) and didn't find a duplicate.
- [X] I've written a clear and descriptive title for this issue, not just "Bug" or "Crash".
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Steps to reproduce

(Edit: Originally I thought it was only affecting the default camera folder but it seems as it has been running for a while now while writing the issue that it is syncing all the folders set up with the first account.)

- Setup InstantUpload for the default camera location for one account on a server, cloud.home.tld
- Setup another account on a different server, cloud.work.tld
- InstantUpload syncs all folders set up to sync with the previously set up account to the newly setup server, cloud.work.tld, despite InstantUpload not being configured for that server.

It might have been triggered by trying to upload two videos to the newly setup server manually but hard to tell because this was the first thing I tried after setting up the second account. (cloud.work.tld). I have had the original account me@cloud.home.tld set up and syncing for months, but recently added the account for the work server to upload two large video files.

### Expected behaviour

InstantUpload to follow sync settings on a server by server & folder by folder setting.

In other words, if **not** set up to sync to one server it should **only** upload files that are uploaded manually.

### Actual behaviour

Nextcloud syncs all the sync folders to the newly setup server (cloud.work.tld) despite it not having the sync setting on.

In the Uploads tab of the newly set up server (cloud.work.tld) I can see the files from all the sync folders being uploaded one by one, however, the user on the second line matches the original server (user@cloud.home.tld) set up with the app.

In the web interface I can confirm that the files have indeed been (re)uploaded to the wrong server.

### Android version

13 (IodeOS 4.0-202301310-FP4)

### Device brand and model

Fairphone 4

### Stock or custom OS?

Custom (explain in "additional information")

### Nextcloud android app version

3.24.1

### Nextcloud server version

25.0.2 

### Using a reverse proxy?

Yes

### Android logs

[logcatOutput.txt](https://github.com/nextcloud/android/files/11547260/logcatOutput.txt)

### Server error logs

```bash
The servers are working fine, receiving files from the app as they should.
```

### Additional information

It seems very similar to, if not the same, bug as [Instant upload with multiple accounts set](https://github.com/nextcloud/android/issues/440)

Luckily I am the admin for both servers so I don't have to worry about the admins of the work server seeing my nudes but would've been quite embarrassing if I wasnt.

**Comment by tribut** [comment]
This has happened for me a lot and it is a very scary bug.

In the App, under "Uploads"/"Uploaded" it shows the correctly configured server "cloud.a.tld" under each of the uploaded files. However, they were not uploaded to "cloud.a.tld", but to "cloud.tld".

I have several Nextcloud servers added to the App, some of which **must not** have access to my private files.

**Comment by nottheend** [comment]
Not sure it is exactly the same issue: For me AutoUpload is partially duplicating it into different folders.

I.e. a local folder is configured to upload to
FolderA/FolderB

But it its acfual behaviour is to upload to
FolderA/FolderB

AND additionally to
FolderC/FolderD/YEAR/MONTH/FolderB

While FolderC/FolderD is a valid upload location for a different local folder

**Comment by eyduh** [comment]
> 

This sounds like a slightly different bug and probably merits its own issue.

**Comment by nottheend** [comment]
Agree, seems to be different.
Although, while searching for a duplicate of the potential bug reported by me, I found two other issue which look like similar to the one in this thread, hence mentioning them here:
#11088 and #11467

**Comment by nottheend** [comment]
created #11755 for the issue above with duplicate auto uploads

**Comment by DetlevCM** [comment]
This also happens with manual uploads.

I sent an email and the bot sent me to Github... Well, I'll add my mails content to this thread (instead of creating a duplicate issue):


> I am using the Nextcloud app to synchronise photos with and upload to my personal Nextcloud instances.
> (I'm using two, one for myself, one for sharing photos.)
> 
> I recently noticed that it is possible for the app to become confused where it is uploading files... (I am adding files manually in this case as I want to share specific files.)
> 
> I.e.
> 
> I create a folder on share/directory - and start uploading photos using the upload files button.
> Possibly as a result of changing accounts in the app, a directory personal/directory is created and files are uploaded there.
> I'm not sure how and why... - It is definitely a result of the app, as the path on the personal instance does not exist and must be created in its entirety.
> 
> Though I get the impression files may have been uploaded to the personal instance on one occasion without switching accounts. I am not sure there...
> The reason I am writing is that I experienced this recently on the 14th.
> The observed behaviour lets me think this is a bug that should be fixed.
> I do not know if this is a new bug or if this has always been present... - Though I didn't experience this in March. End of May/June would be the earliest, and it seems to have happened only twice, maybe thrice thus far.

**Comment by DetlevCM** [comment]
As complete speculation without looking at the code, I wonder if something along the following lines happens:
 - a file is added to the upload queue (manually or automatically) which contains the source file and the target path
 - the nextcloud path uploads to the currently selected account 
 
This would sort of explain the behaviour - and once an upload is started on an account, it is finshed on that account. 
Though that might be not the entire story as I had some large files uploaded that even with 5G would take more time... Maybe there is some kind of batching going on?

I would speculate that the files in the upload queue are not explicitly bound to a specific account when they are added to the upload queue: IF this is the case, this could be an easy fix (tie files in upload to an account).
But I leave this to people with actual experience with the code...

**Comment by Corvus4n6** [comment]
Same findings and very problematic. There needs to be strict separation of configurations between accounts/instances.

I was testing the auto-upload feature on photos where I had multiple account configured in my Nextcloud app. The setting was made to auto upload photos to account 1, but the photos were uploaded to *both* account 1 and account 2. This is a rather significant security and privacy issue. If someone is using multiple instances and intends to upload to their personal, they will find their data *also* uploaded to another instance at the same time.

**Comment by nottheend** [comment]
Is there a label like "critical" in terms of data sensitivity which could be assigned to this issue?

**Comment by coyoteseattle** [comment]
I've been running into what appears to be this same issue.  I also have two nextcloud servers configured in the android app, and have configured it to instant upload pictures to both, for redundancy.  On one server, the app is configured to sync the camera to /Photos, and on the other, it was configured to sync to /InstantUpload.  
When an upload is triggered, the first server receives HEAD requests from the app for the file under both /Photos AND /InstantUpload, but then nothing else.  The app then does the normal upload sequence on the second server, for both paths.
To attempt a workaround, I changed the second server's configuration to also sync to /Photos, instead of /InstantUpload.  It is currently trying to upload each file to the first server twice with the same path, but is otherwise working as expected.

**Comment by joshtrichards** [comment]
Closing in favor of consolidating tracking in #11088

**Comment by joshtrichards** [comment]
Fixed in #12516
